SQL Server PKINIT authentication update

gabrielcorado opened the pull request:
Related to #13837.

This PR adds the missing configuration options, bug fixes, and improvements regarding the SQL Server PKINIT authentication method.

Here is an example of what the configuration database services would look like:
```yaml
db_service:
  enabled: "yes"
  databases:
    - name: sqlserver
      protocol: sqlserver
      uri: MYDB.gabrielcorado.teleport.dev:1433
      tls: # If necessary
        mode: verify-ca
        ca_cert_file: "custom-ca.pem"
      ad:
        spn: MSSQLSvc/MYDB.gabrielcorado.teleport.dev:1433
        kdc_host_name: DOMAINCONTROLLER.gabrielcorado.teleport.dev
        domain: gabrielcorado.teleport.dev
        ldap_cert: |
          -----BEGIN CERTIFICATE-----
          ...
          -----END CERTIFICATE-----
```

And how to configure the database access on Windows Active Directory:
<details><summary>Instructions</summary>

1. Generate Database CA and CRL in Windows format (introduced at #22578):
    1. `tctl auth export --type=db-der > db-ca.crt`
    2. `tctl auth crl --type=db > db-ca.crl`
2. Create GPO Policy to "Teleport DB Access"
    1. Create GPO and Link.
    3. Edit GPO:
		1. Go to `Policies -> Windows Settings -> Security Settings -> System Services -> Smart Card`: Check "Automatic".
		2. Go to `Policies -> Windows Settings -> Security Settings -> Public Key Policies -> Trusted Root Certification Authorities`: Click on import and import Database CA cert.
		4. Force policies update: `gpupdate.exe /force`
3. Import Teleport CA and CRL into AD Store:
	* `certutil -dspublish -f db-ca.crt RootCA`
	* `certutil -dspublish -f db-ca.crt NTAuthCA`
	* `certutil -dspublish -f db-ca.crl Teleport`
	* `certutil -pulse`
4. Export LDAP CA and copy it into the configuration.
    * [An example of how to export it using PowerShell](https://github.com/gravitational/teleport/blob/master/lib/web/scripts/desktop/configure-ad.ps1#L127-L136)
</details>

gabrielcorado committed:
feat: fix and improve kinit sqlserver

Tener reviewed (commented):
`TestKRBConfString` leaves behind `lib/srv/db/sqlserver/kinit/krb5_cache/alice@example.com`; we should use temp dir for that file instead.

gabrielcorado commented on lib/srv/db/sqlserver/kinit/kinit.go:
```diff
@@ -281,59 +285,65 @@ func (k *CommandLineInitializer) UseOrCreateCredentials(ctx context.Context) (*c
+		return nil, nil, trace.Wrap(err)
+	}
+
+	conf, err := config.Load(krbConfPath)
```
We're just returning the `conf`. I've made this change so that we don't need to rely on a `krb5.conf` file provided by the users. Instead, we create the `*config.Config` struct based on the KRB file generated by this function (`UseOrCreateCredentials`).

Tener commented on lib/services/database.go:
```diff
@@ -182,8 +182,8 @@ func ValidateDatabase(db types.Database) error {
-		if db.GetAD().KeytabFile == "" {
-			return trace.BadParameter("missing keytab file path for database %q", db.GetName())
+		if db.GetAD().KeytabFile == "" && db.GetAD().KDCHostName == "" {
+			return trace.BadParameter("missing keytab file path or kdc_host_name for database %q", db.GetName())
```
Ah, I see your point. The error may be ambiguous as it is now. What do you think about:
```suggestion
			return trace.BadParameter("either keytab file path or kdc_host_name must be provided for database %q, both are missing", db.GetName())
```

gabrielcorado committed:
test(kinit): use a temp dir for writing ccache files

gabrielcorado committed:
chore(services): update validation error message

gabrielcorado committed:
chore(kinit): improve field description

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/pkinit-sqlserver

smallinsky reviewed (commented):
@gabrielcorado 
Sorry for the review delay and thanks for docs that was relay helpful bootstrapping test env in order to test this PKINIT functionality . 

I have encounter a few issue following the docs guides: 
- getting databaseCA from db agent with buildIn role: access denied error 
- tctl auth crl --type=db > db-ca.crl Premission error. The endpoint can be called only with buildIn role.  

but with hardcoded workarounds in code I  was finally able to run successfully the full flow. 
This code change LGTM but could you test one more time the full flow from scratch and ensure that it runs successfully ?

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/pkinit-sqlserver

gabrielcorado committed:
refactor(kinit): use CA from certificate generation

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/pkinit-sqlserver

gabrielcorado commented on lib/srv/db/sqlserver/kinit/kinit.go:
```diff
@@ -281,59 +287,65 @@ func (k *CommandLineInitializer) UseOrCreateCredentials(ctx context.Context) (*c
 	wca, err := k.certGetter.GetCertificateBytes(ctx)
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return nil, nil, trace.Wrap(err)
```
Not sure if we can use it as it will contain the DB CA (specified by the user on the configuration). However, [I updated the implementation](https://github.com/gravitational/teleport/pull/22792/commits/2efaf5743c9a5d68160af4782e6a2e6051b3b831) to use the CA returned from the certificate generation as it seems more accurate since we need the CA that generated the certificate.

gabrielcorado commented:
@smallinsky Thanks for testing it.

> * tctl auth crl --type=db > db-ca.crl Premission error. The endpoint can be called only with buildIn role.

I've pushed a separate PR to fix command #23327.

gabrielcorado committed:
refactor(kinit): remove unused function

smallinsky reviewed (commented):
Why do we need `krb5_cache/alice@example.com`  file in this PR. There is already kinit/testdata dir  ? 
Aport from that LGTM.

gabrielcorado committed:
chore(kinit): remove leftover file

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/pkinit-sqlserver

gabrielcorado commented:
> Why do we need `krb5_cache/alice@example.com` file in this PR. There is already kinit/testdata dir ?

It was generated by the test and got committed accidentally; I've changed the test to use a temporary directory and dropped the file.

greedy52 commented on lib/srv/db/sqlserver/kinit/kinit.go:
```diff
@@ -361,3 +338,10 @@ func (k *CommandLineInitializer) WriteKRB5Config(path string) error {
+// The file must contain the Teleport DB CA and the KDB/LDAP CA, otherwise the
+// connections will fail.
+func (k *CommandLineInitializer) buildAnchorsFileContents(caBytes []byte) []byte {
+	return append(caBytes, []byte(k.ldapCertificatePEM)...)
```
do we need a new line?

greedy52 replied:
If we are sure it has a new line then it's ok. If we are not sure, an extra newline wouldn't hurt.

greedy52 commented on lib/srv/db/sqlserver/kinit/kinit_test.go:
```diff
@@ -170,10 +175,11 @@ func TestKRBConfString(t *testing.T) {
 			User:        "alice",
 			Realm:       "example.com",
-			KDCHost:     "host.example.com",
+			KDCHost:     "instance.host.example.com",
```
Can `KDCHost` be empty?

gabrielcorado replied:
No, users should provide it, even if it is the same as their domain.

greedy52 replied:
Is `KDCHost` passed in from config as `kdc_host_name` now? For older versions, it would be empty?

gabrielcorado replied:
> Is KDCHost passed in from config as kdc_host_name now? For older versions, it would be empty?


Yes. In older versions, its value would always be empty. [But PKINIT won't be used if there is no KDCHost and LDAPCert.](https://github.com/gravitational/teleport/blob/master/lib/srv/db/sqlserver/connect.go#L66)

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/pkinit-sqlserver

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/pkinit-sqlserver
